File: backend/src/models/task.py
from enum import Enum
from sqlmodel import SQLModel, Field, Column, DateTime, String, Relationship
from datetime import datetime
from typing import Optional, List, TYPE_CHECKING

# Import TaskTag for link_model (must be actual class, not string)
from .tag import TaskTag

if TYPE_CHECKING:
    from .tag import Tag, TagRead


class Priority(str, Enum):
    """Task priority levels with sort order: urgent(0) > high(1) > medium(2) > low(3)"""
    low = "low"
    medium = "medium"
    high = "high"
    urgent = "urgent"


class RecurrenceRule(str, Enum):
    """Recurrence patterns for recurring tasks"""
    daily = "daily"
    weekly = "weekly"
    monthly = "monthly"
    custom = "custom"  # Uses recurrence_interval for custom day count

class TaskBase(SQLModel):
    """Base fields shared across Task schemas"""
    title: str
    description: Optional[str] = None
    completed: bool = False
    priority: Priority = Priority.medium
    due_date: Optional[datetime] = None
    reminder_at: Optional[datetime] = None
    is_recurring: bool = False
    recurrence_rule: Optional[RecurrenceRule] = None
    recurrence_interval: Optional[int] = None


class Task(TaskBase, table=True):
    """Task database model with all Phase VI fields"""
    __tablename__ = "tasks"

    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: str = Field(sa_column=Column(String, nullable=False, index=True))
    reminder_sent: bool = Field(default=False)
    parent_task_id: Optional[int] = Field(default=None, foreign_key="tasks.id")

    created_at: datetime = Field(
        sa_column=Column(DateTime(timezone=True), nullable=False),
        default_factory=datetime.utcnow
    )
    updated_at: datetime = Field(
        sa_column=Column(DateTime(timezone=True), nullable=False),
        default_factory=datetime.utcnow
    )

    # Relationships
    tags: List["Tag"] = Relationship(
        back_populates="tasks",
        link_model=TaskTag
    )

class TaskCreate(TaskBase):
    """Schema for creating a new task"""
    tag_ids: List[int] = Field(default_factory=list)


class TaskRead(TaskBase):
    """Schema for reading a task with all fields"""
    id: int
    user_id: str
    reminder_sent: bool
    parent_task_id: Optional[int]
    created_at: datetime
    updated_at: datetime
    tags: List["TagRead"] = Field(default_factory=list)


class TaskUpdate(SQLModel):
    """Schema for updating a task - all fields optional"""
    title: Optional[str] = None
    description: Optional[str] = None
    completed: Optional[bool] = None
    priority: Optional[Priority] = None
    due_date: Optional[datetime] = None
    reminder_at: Optional[datetime] = None
    is_recurring: Optional[bool] = None
    recurrence_rule: Optional[RecurrenceRule] = None
    recurrence_interval: Optional[int] = None
    tag_ids: Optional[List[int]] = None

File: frontend/src/types/task.ts
// Task Types for Advanced Todo Features (Phase VI)

export type Priority = "low" | "medium" | "high" | "urgent";

export type RecurrenceRule = "daily" | "weekly" | "monthly" | "custom";

export interface Tag {
  id: number;
  name: string;
  color: string;
  user_id: string;
  created_at: string;
}

export interface TagCreate {
  name: string;
  color?: string;
}

export interface TagUpdate {
  name?: string;
  color?: string;
}

export interface Task {
  id: number;
  title: string;
  description: string | null;
  completed: boolean;
  user_id: string;
  priority: Priority;
  due_date: string | null;
  reminder_at: string | null;
  reminder_sent: boolean;
  is_recurring: boolean;
  recurrence_rule: RecurrenceRule | null;
  recurrence_interval: number | null;
  parent_task_id: number | null;
  created_at: string;
  updated_at: string;
  tags: Tag[];
}

export interface TaskCreate {
  title: string;
  description?: string;
  priority?: Priority;
  due_date?: string;
  reminder_at?: string;
  is_recurring?: boolean;
  recurrence_rule?: RecurrenceRule;
  recurrence_interval?: number;
  tag_ids?: number[];
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  completed?: boolean;
  priority?: Priority;
  due_date?: string | null;
  reminder_at?: string | null;
  is_recurring?: boolean;
  recurrence_rule?: RecurrenceRule | null;
  recurrence_interval?: number | null;
  tag_ids?: number[];
}

// Response type for completing recurring tasks
export interface RecurringTaskCompleteResponse {
  completed_task: Task;
  next_task?: Task;
}

// Filter parameters for GET /api/tasks
export interface TaskFilters {
  search?: string;
  status?: "all" | "pending" | "completed";
  priority?: Priority[];
  tags?: number[];
  due_before?: string;
  due_after?: string;
  overdue?: boolean;
  sort_by?: "created_at" | "due_date" | "priority" | "title";
  order?: "asc" | "desc";
}

// Priority color mapping for UI
export const PRIORITY_COLORS: Record<Priority, string> = {
  low: "#9CA3AF",      // gray-400
  medium: "#3B82F6",   // blue-500
  high: "#F97316",     // orange-500
  urgent: "#EF4444",   // red-500
};

// Priority sort order for sorting (lower = higher priority)
export const PRIORITY_ORDER: Record<Priority, number> = {
  urgent: 0,
  high: 1,
  medium: 2,
  low: 3,
};
